### ISSUE TITLE
bad file descriptor error

### ISSUE THREAD

**Issue by jacksu** [oktal/pistache#501]
![QQ截图20190318213901](https://user-images.githubusercontent.com/6895190/54533918-40bd2280-49c6-11e9-9d8a-0b023f208fc0.png)

**Comment by jacksu** [comment]
https://github.com/oktal/pistache/blob/master/src/common/os.cc#L187

**Comment by knowledge4igor** [comment]
@jacksu Do you have this problem on latest source? What are the steps for reproducing this uncaught exception?

**Comment by albertochm95** [comment]
I have the same problem. I updated pistache and the problem continues

**Comment by dmorton-spirent** [comment]
My recent PR #536 added a unit test that triggers this error message. It made the build flaky and I spent a day digging further into it. Results of my investigation are included in this comment: 
https://github.com/oktal/pistache/pull/536#issuecomment-491460208.

tl;dr version: if the HTTP parser is called more than once and the first call contains text that causes a parser exception to be thrown (i.e. invalid header value) and the second call contains a HTTP header (even a valid one), then the above error is thrown when the second call tries to send back a second error to the client.

**Comment by dmorton-spirent** [comment]
Further investigation has, I believe, uncovered the root cause. Continuing on from the thread above. When the first HTTP parser exception is thrown Pistache catches it and sends an error response to the client. The code that handles transmitting the error response defaults to closing the connection after transmitting the reply. Connection is kept open if and only if the request contained the `Connection: Keep-Alive` header. 
https://github.com/oktal/pistache/blob/4355267e20a912992006b2ca914147a8cb49c3fa/src/common/http.cc#L651-L655
and
https://github.com/oktal/pistache/blob/4355267e20a912992006b2ca914147a8cb49c3fa/src/common/http.cc#L689-L692
In the error condition from the unit test the `Connection: Keep-Alive` header is either not sent at all or sent after the invalid header.

When Pistache tries to send the second error reply the `send()` call returns `-1` on account of the file descriptor being closed. The current error checks don't check for `EBADF` in that case. Instead, control passes to the default error path which discards the reply and reconfigures the file descriptor. It's this reconfigure step that causes the crash.

https://github.com/oktal/pistache/blob/4355267e20a912992006b2ca914147a8cb49c3fa/src/common/transport.cc#L262-L275
